### PR TITLE
Bluetooth: L2CAP: remove `l2cap_chan_create_seg()`

### PR DESCRIPTION
**Pull request by jori-nordic**
Remove the logic for adding the SDU length when allocating segments. That
section was dead code after the recent patches.

Inline the remainder of the logic into `l2cap_chan_le_send()`.

Stacked on https://github.com/zephyrproject-rtos/zephyr/pull/67528

### CODE REVIEW

File: subsys/bluetooth/host/l2cap.c [hunk 1]
```diff
@@ -1996,16 +1947,35 @@ static int l2cap_chan_le_send(struct bt_l2cap_le_chan *ch,
...
 	/* Save state so it can be restored if we failed to send */
 	net_buf_simple_save(&buf->b, &state);
 
-	seg = l2cap_chan_create_seg(ch, buf, sdu_hdr_len);
-	if (!seg) {
-		atomic_inc(&ch->tx.credits);
-		return -EAGAIN;
+	if (net_buf_frags_len(buf) <= ch->tx.mps) {
+		LOG_DBG("len < MPS, not allocating seg for %p", buf);
+		seg = net_buf_ref(buf);
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/l2cap.c:1959]: This look suspicious to me. I think we should move the reference from `buf` to `seg`, not copy it. Can you double-check this?

**Review by alwa-nordic — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/l2cap.c (hunk 1 above)
Comment (jori-nordic) [reply subsys/bluetooth/host/l2cap.c:1959]: let's look at it IRL today

File: subsys/bluetooth/host/l2cap.c [hunk 2]
```diff
@@ -1996,16 +1947,35 @@ static int l2cap_chan_le_send(struct bt_l2cap_le_chan *ch,
...
+
+		len = seg->len;
+	} else {
+		LOG_DBG("allocating segment for %p (%u bytes left)", buf, buf->len);
+		seg = l2cap_alloc_seg(buf, ch);
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/l2cap.c:1964]: Remove `buf` argument. It's dead code.

File: subsys/bluetooth/host/l2cap.c (hunk 1 above)
Comment (alwa-nordic) [reply subsys/bluetooth/host/l2cap.c:1959]: Looks ok. `buf` is not moving a reference. It's a borrow.

**Review by alwa-nordic — CHANGES_REQUESTED** [review]
See comments

File: subsys/bluetooth/host/l2cap.c [hunk 3]
```diff
@@ -2002,10 +1953,29 @@ static int l2cap_chan_le_send(struct bt_l2cap_le_chan *ch,
 	/* Save state so it can be restored if we failed to send */
 	net_buf_simple_save(&buf->b, &state);
 
-	seg = l2cap_chan_create_seg(ch, buf, sdu_hdr_len);
-	if (!seg) {
-		atomic_inc(&ch->tx.credits);
-		return -EAGAIN;
+	if (net_buf_frags_len(buf) <= ch->tx.mps) {
+		LOG_DBG("len < MPS, not allocating seg for %p", buf);
```
Comment (Thalley) [inline subsys/bluetooth/host/l2cap.c:1957]:
```suggestion
	if (net_buf_frags_len(buf) <= ch->tx.mps) {
		LOG_DBG("len <= MPS, not allocating seg for %p", buf);
```

Small mismatch between code and comment

**Review by Thalley — DISMISSED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Review by Thalley — APPROVED** [review]

**Comment by carlescufi** [comment]
@jhedberg can you take a look?

**Review by jhedberg — APPROVED** [review]

**Pull request merged**